app/code/Magento/InventoryCatalog/Model/UpdateLegacyStockItemByDefaultSourceItem.php

```diff
@@ -0,0 +1,88 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+declare(strict_types=1);
+
+namespace Magento\InventoryCatalog\Model;
+
+use Magento\Catalog\Model\ProductIdLocatorInterface;
+use Magento\CatalogInventory\Api\Data\StockItemInterface;
+use Magento\Framework\App\ResourceConnection;
+use Magento\Framework\Exception\NoSuchEntityException;
+use Magento\InventoryApi\Api\Data\SourceItemInterface;
+use Magento\InventoryCatalog\Api\DefaultSourceProviderInterface;
+use Magento\InventoryCatalog\Api\DefaultStockProviderInterface;
+
+/**
+ * Legacy update cataloginventory_stock_item by plain MySql query.
+ * Use for skip save by \Magento\CatalogInventory\Model\ResourceModel\Stock\Item::save
+ */
+class UpdateLegacyStockItemByDefaultSourceItem
+{
+    /**
+     * @var ResourceConnection
+     */
+    private $resourceConnection;
+
+    /**
+     * @var DefaultSourceProviderInterface
+     */
+    private $defaultSourceProvider;
+
+    /**
+     * @var ProductIdLocatorInterface
+     */
+    private $productIdLocator;
+    /**
+     * @var DefaultStockProviderInterface
+     */
+    private $defaultStockProvider;
+
+    /**
+     * @param ResourceConnection $resourceConnection
+     * @param DefaultSourceProviderInterface $defaultSourceProvider
+     * @param DefaultStockProviderInterface $defaultStockProvider
+     * @param ProductIdLocatorInterface $productIdLocator
+     */
+    public function __construct(
+        ResourceConnection $resourceConnection,
+        DefaultSourceProviderInterface $defaultSourceProvider,
+        DefaultStockProviderInterface $defaultStockProvider,
+        ProductIdLocatorInterface $productIdLocator
+    ) {
+        $this->resourceConnection = $resourceConnection;
+        $this->defaultSourceProvider = $defaultSourceProvider;
+        $this->defaultStockProvider = $defaultStockProvider;
+        $this->productIdLocator = $productIdLocator;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function execute(SourceItemInterface $sourceItem)
+    {
+        if ($sourceItem->getSourceId() != $this->defaultSourceProvider->getId()) {
+            return;
+        }
+
+        $productIds = $this->productIdLocator->retrieveProductIdsBySkus([$sourceItem->getSku()]);
+        $productId = isset($productIds[$sourceItem->getSku()]) ? key($productIds[$sourceItem->getSku()]) : false;
+
+        if ($productId) {
+            $connection = $this->resourceConnection->getConnection();
+            $connection->update(
+                $this->resourceConnection->getTableName('cataloginventory_stock_item'),
+                [
+                    StockItemInterface::QTY => $sourceItem->getQuantity(),
+                ],
+                [
+                    StockItemInterface::STOCK_ID . ' = ?' => $this->defaultStockProvider->getId(),
+                    StockItemInterface::PRODUCT_ID . ' = ?' => $productId,
+                    'website_id = ?' => 0
```

Comment:
Need to check how do we write website value in this table
Looks like we need to use some service for resolving website id (instead of hardcoded value)